dataikuapi/dss/ml.py

```diff
@@ -534,41 +551,189 @@ def __init__(self, name, algo_settings):
+                        " expecting a non-empty list of numbers"
+        if values is not None:
+            assert isinstance(values, list) and len(values) > 0, error_message
+            for val in values:
+                assert isinstance(val, int) or isinstance(val, float), error_message
+            limit_min = self._algo_settings[self.name]["limit"].get("min")
+            if limit_min is not None:
+                assert all(limit_min <= val for val in values), "Value(s) below hyperparameter \"{}\" limit {}".format(self.name, limit_min)
+            limit_max = self._algo_settings[self.name]["limit"].get("max")
+            if limit_max is not None:
+                assert all(val <= limit_max for val in values), "Value(s) above hyperparameter \"{}\" limit {}".format(self.name, limit_max)
+            if len(set(values)) < len(values):
+                warnings.warn("Detected duplicates in provided values: " + str(sorted(values)))
+            self._algo_settings[self.name]["values"] = values
+        if set_mode_to_explicit:
+            self._set_mode_to_explicit()
+
+    @property
+    def values(self):
+        return self._algo_settings[self.name]["values"]
+
+    @values.setter
+    def values(self, values):
+        self.set_explicit_values(values)
+
+    def _check_number_input(self, input):
+        assert isinstance(input, int) or isinstance(input, float), \
+            "Invalid input type for hyperparameter \"{}\": ".format(self.name) + \
+            "range bounds must be numbers"
 
-    def set_mode_to_range(self, range_min=None, range_max=None, nb_values=None):
-        self._algo_settings._set_numerical_range(self.name, range_min=range_min, range_max=range_max, nb_values=nb_values)
+    def set_range(self, min=None, max=None, nb_values=None, set_mode_to_range=True):
+        if min is None and max is None and nb_values is None:
+            warnings.warn("Numerical range for hyperparameter \"{}\" not modified".format(self.name))
+        else:
+            if min is not None:
+                self._check_number_input(min)
+                limit_min = self._algo_settings[self.name]["limit"].get("min")
+                if limit_min is not None:
+                    assert limit_min <= min, "Range min {} is below hyperparameter \"{}\" limit {}".format(min, self.name, limit_min)
+            if max is not None:
+                self._check_number_input(max)
+                limit_max = self._algo_settings[self.name]["limit"].get("max")
+                if limit_max is not None:
+                    assert max <= limit_max, "Range max {} is above hyperparameter \"{}\" limit {}".format(max, self.name, limit_max)
+            if nb_values is not None:
+                assert isinstance(nb_values, int) and nb_values >= 2, "Range number of values for hyperparameter \"{}\" must be an integer and >= 2".format(self.name)
+
+            if min is not None:
+                self._algo_settings[self.name]["range"]["min"] = min
```

Comment:
Actually, I don't understand what you are suggesting: 
- all setters of `Range` end up calling this method, so this would mean swapping code snippets between `NumericalHyperparameterSettings._set_range` and the setters of `Range` and I'd rather keep the complexity in one entrypoint (`_set_range`)
- we don't want to change the type of `self._algo_settings[self.name]["range"]` (a `dict`).

Reply:
Actually forget it. I thought the `Range` class would contain all the methods to modify the range (a bit like the various `HyperparameterSettings` classes contain the methods to modify the hyperparams settings).